web/js/codeworld_shared.js

```diff
@@ -1559,78 +1559,84 @@ function preFormatMessage(msg) {
 }
 
 function printMessage(type, message) {
-  const outputDiv = document.getElementById('message');
+  const $outputBlock = $('#message');
 
-  let box = outputDiv.lastChild;
-  let messageContent;
-  if (box && type === 'log' && box.classList.contains('log')) {
-    box.rawMessage += message;
-    messageContent = box.lastChild;
-  } else {
-    box = document.createElement('div');
-    box.classList.add('message-box');
-    box.classList.add(type);
+  let $box = $outputBlock.find('.message-box');
+  let $messageContent = $outputBlock.find('.message-content');
 
-    const messageGutter = document.createElement('div');
-    messageGutter.classList.add('message-gutter');
+  if (!$box.length && !$messageContent.length) {
```

Comment:
I think the logic is wrong here.  If the type is 'log', and the last output type is also 'log', then this should be appended.  Otherwise, it should be added as a new box.

In particular:
1. If a 'log' message comes after an 'error' or 'warning' message, it should be in a new box.
2. If there are two 'error's or 'warning's in a row, they should still be in separate boxes.
3. ONLY if there's two 'log's in a row should they be combined.

Reply:
I haven't considered this at all. Will correct the logic, thanks for clarifying!